examples/source_examples/newsapi/connector.py

```diff
@@ -0,0 +1,197 @@
+# The function takes two parameters:
+# - configuration: dictionary contains any secrets or payloads you configure when deploying the connector
+# - state: a dictionary contains whatever state you have chosen to checkpoint during the prior sync
+# The state dictionary is empty for the first sync or for any full re-sync
+def update(configuration: dict, state: dict):
+
+    try:
+        conf = configuration
+        base_url = "https://newsapi.org/v2/everything"
+        from_ts = state['to_ts'] if 'to_ts' in state else '2024-11-10T00:00:00'
+        now = datetime.datetime.now()
+        to_ts = now.strftime("%Y-%m-%dT%H:%M:%S")
+        headers = {"Authorization": "Bearer "+conf["API_KEY"], "accept": "application/json"}
+
+        params = \
+            {"from": from_ts,
+            "to": to_ts,
+            "page": "1",
+            "language": "en",
+             "sortBy": "publishedAt",
+             "pageSize": conf["pageSize"]}
+
+        topics = json.loads(conf["topic"])
+
+        for t in topics:
+            params["q"] = t
+            params["page"] = "1"
+            yield from sync_items(base_url, headers, params, state, t)
+
+
+        # Update the state with the new cursor position, incremented by 1.
+        new_state = {
+            "to_ts": to_ts
+        }
+        log.fine(f"state updated, new state: {repr(new_state)}")
+
+        # Yield a checkpoint operation to save the new state.
+        yield op.checkpoint(state=new_state)
+
+    except Exception as e:
+        # Return error response
+        exception_message = str(e)
+        stack_trace = traceback.format_exc()
+        detailed_message = f"Error Message: {exception_message}\nStack Trace:\n{stack_trace}"
+        raise RuntimeError(detailed_message)
+
+
+# The function takes five parameters:
+# - base_url: The URL to the API endpoint.
+# - headers: Authentication headers
+# - state: State dictionary
+# - topic: current topic to search
+# - params: A dictionary of query parameters to be sent with the API request.
+def sync_items(base_url, headers, params, state, topic):
+    more_data = True
+
+    while more_data:
+        # Get response from API call.
+        response_page = get_api_response(base_url, headers, params)
+
+        log.info(str(response_page["totalResults"]) + " results for topic " + topic)
+
+        # Process the items.
+        items = response_page.get("articles", [])
+        if not items:
+            break  # End pagination if there are no records in response.
+
+        # Iterate over each user in the 'items' list and yield an upsert operation.
+        # The 'upsert' operation inserts the data into the destination.
+        # Update the state with the 'updatedAt' timestamp of the current item.
+        summary_first_item = {'title': items[0]['title'], 'source': items[0]['source']}
+        log.info(f"processing page of items. First item starts: {summary_first_item}, Total items: {len(items)}")
```

Comment:
Logging every page is fine. But if page size is small it can create too many logs. Should we add a note or remove this log?

Reply:
I'm fine with removing.